rfcs/2020-08-26-3191-host-metrics.md

```diff
@@ -0,0 +1,157 @@
+
+## Scope
+
+This RFC will cover:
+
+- A new source for host-based metrics, specifically:
+  - CPU
+  - Memory
+  - Disk
+- Collection on Linux, Windows, OSX.
+
+This RFC will not cover:
+
+- Other host metrics.
+- Other platforms
+
+## Motivation
+
+Users want to collect, transform, and forward metrics to better observe how their hosts are performing.
+
+## Internal Proposal
+
+Build a single source called `host_metrics` (name to be confirmed) to collect host/system level metrics.
+
+I've found a number of possible Rust-based solutions for implementing this collection, cross-platform.
+
+- https://crates.io/crates/heim (Rust)
+- https://lib.rs/crates/sys-info (Rust + C)
+- https://github.com/myfreeweb/systemstat (pure Rust)
+- https://docs.rs/sysinfo/0.3.19/sysinfo/index.html
+
+The Heim crate has a useful comparison doc showing the available tools and their capabilities.
+
+- https://github.com/heim-rs/heim/blob/master/COMPARISON.md
+
+For this implementation we're recommending the Heim crate based on platform and feature coverage.
+
+We'd use one of these to collect the following metrics:
+
+- `cpu_seconds_total` by mode (idle, nice, system, user) per CPU. (counter)
+- `disk_read_bytes_total` by disk (counter)
+- `disk_read_errors_total` by disk (counter)
+- `disk_read_retries_total` by disk (counter)
+- `disk_read_sectors_total` by disk (counter)
+- `disk_read_time_seconds_total` by disk (counter)
+- `disk_reads_completed_total` by disk (counter)
+- `disk_write_errors_total` by disk (counter)
+- `disk_write_retries_total` by disk (counter)
+- `disk_write_time_seconds_total` by disk (counter)
+- `disk_writes_completed_total` by disk (counter)
+- `disk_written_bytes_total` by disk (counter)
+- `disk_written_sectors_total` by disk (counter)
+- `filesystem_avail_bytes` labeled with device, filesystem type, and mountpoint (gauge)
+- `filesystem_device_error` labeled with device, filesystem type, and mountpoint (gauge)
+- `filesystem_total_file_nodes` labeled with device, filesystem type, and mountpoint (gauge)
+- `filesystem_free_file_nodes` labeled with device, filesystem type, and mountpoint (gauge)
+- `filesystem_free_bytes` labeled with device, filesystem type, and mountpoint (gauge)
+- `filesystem_size_bytes` labeled with device, filesystem type, and mountpoint (gauge)
+- `load1` (gauge)
+- `load5` (gauge)
+- `load15` (gauge)
+- `memory_active_bytes` (gauge)
+- `memory_compressed_bytes` (gauge)
+- `memory_free_bytes` (gauge)
+- `memory_inactive_bytes` (gauge)
+- `memory_swap_total_bytes` (gauge)
+- `memory_swap_used_bytes` (gauge)
+- `memory_swapped_in_bytes_total` (gauge)
+- `memory_swapped_out_bytes_total` (gauge)
+- `memory_total_bytes` (gauge)
+- `memory_wired_bytes` (gauge)
+
+Users should be able to limit the collection of metrics to specific classes, here: `cpu`, `memory`, `disk/filesystem`, `load`.
+
+Metrics will also be labeled with:
+
+- `host`: the host name of the host being monitored.
```

Comment:
I think we'll probably need some more labels for some of the metrics to be useful:

* `device` for disk based metrics
* `cpu` for CPU based metrics
* `filesystem` for filesystem based metrics

Reply:
Sorry, I'm realizing I was unclear. I think you've addressed this above though with the update to the metrics list. I had meant, for example, that disk based metrics should be labeled (or tagged; not sure which term we prefer) with the device the metric is associated with (e.g. `device=/dev/sda`).

I think I like these "collector" labels too though. I think they'd be better as simply `collector` so you'd have things like `collector=disk`.

Reply:
Ah. Yeah - this is a terminology mismatch. I'll update to say tagged and make it clearer.